Coin Selection Specification: Draft 1

jonathanknowles opened the pull request:
## Related Issue

https://github.com/input-output-hk/cardano-coin-selection/issues/21

## Summary

This PR contains a draft of a specification for the coin selection algorithms used by Cardano Wallet.

The purpose of the specification is to provide an easy-to-understand, **_human-readable_** description of these algorithms that doesn't require an understanding of Haskell.

The purpose of this PR is to gain feedback on areas for improvement. Please feel free to leave comments and constructive suggestions!

The document is written in Markdown, and can be read in its [rendered form](https://github.com/input-output-hk/cardano-coin-selection/blob/jonathanknowles/specification-draft-1/information/specification.md).

Please note that there are several **areas not addressed by this specification**:

* There is no description of the algorithm used for **migration** of funds from one wallet to another.
* Descriptions of **fee balancing** algorithms are not included.

jonathanknowles committed:
Draft 1 of the coin selection specification.

jonathanknowles commented on information/specification.md:
```diff
@@ -0,0 +1,705 @@
+Addresses are owned (and generated) by individual wallets.
+
+Coin selection algorithms are, in general, agnostic to the details of
+individual address formats. Therefore, such formats will not be discussed here.
```
```suggestion
In general, coin selection algorithms are agnostic to the choice of address
format. Therefore, the details of address formats are not described within this
document.
```

jonathanknowles commented on information/specification.md:
```diff
@@ -0,0 +1,705 @@
+    This failure occurs when the total value of the entries within the [initial
+    UTxO set](#initial-utxo-set) (the amount of money _available_) is _less
+    than_ the the total value of all entries in the [initial output
+    list](#requested-output-list) (the amount of money _required_).
```
```suggestion
    than_ the the total value of all entries in the [requested output
    list](#requested-output-list) (the amount of money _required_).
```

jonathanknowles commented on information/specification.md:
```diff
@@ -0,0 +1,705 @@
+    [requested output list](#requested-output-list).
+
+    All algorithms require that there is _at least one_ UTxO entry available
+    _for each_ output.
```
```suggestion
    _for each_ requested output.
```

jonathanknowles commented on information/specification.md:
```diff
@@ -0,0 +1,705 @@
+  * [Random-Improve](#random-improve)
+
+In general, Cardano Wallet gives _priority_ to the
+[Random-Improve](#random-improve) algorithm.
```
Perhaps add a short paragraph here explaining *why* Cardano Wallet gives priority to Random-Improve. This paragraph could link to the "Self Organization in Coin Selection" blog article (listed at the end as an external resource).

jonathanknowles replied:
@jonathanknowles wrote:
> 

Perhaps add a short paragraph here explaining why Cardano Wallet gives priority to Random-Improve. This paragraph could link to the "Self Organization in Coin Selection" blog article (listed at the end as an external resource).

Fixed in ebfd135cbbfe2bf7aff749d326f30deca462397f.

jonathanknowles commented on information/specification.md:
```diff
@@ -0,0 +1,705 @@
+  * processes outputs in _ascending order of coin value_.
+
+  * continues to select values from the [available UTxO
+    set](#available-utxo-set)
```
```suggestion
    set](#available-utxo-set).
```

jonathanknowles commented on information/specification.md:
```diff
@@ -0,0 +1,705 @@
+ 1.  **Calculates a _target range_** for the total value of inputs used to
+     pay for that output, defined by the triplet:
+
+     (_minimum_, _ideal_, _maximum_) = (_v_, _2v_, _2v_)
```
This is a typo. The _maximum_ should be _3v_.

```suggestion
     (_minimum_, _ideal_, _maximum_) = (_v_, _2v_, _3v_)
```

jonathanknowles commented on information/specification.md:
```diff
@@ -0,0 +1,705 @@
+        Count](#maximum-input-count).
+
+ 3.  **Creates a _change value_** for the output, equal to the total value
+     of the _final UTxO selection_ for that output minus the value _v_ of
```
Consider a change in terminology here:
```suggestion
     of the _improved UTxO selection_ for that output minus the value _v_ of
```

jonathanknowles committed:
Add missing full stop.

Fixes https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r396242947

jonathanknowles committed:
Use correct coefficient for target maximum.

The maximum should be 3v, rather than 2v.

Fixes https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r396243248

jonathanknowles committed:
Revise paragraph relating to address format agnosticism.

This change replaces "here" with the more specific "this document", and clarifies that the agnosticism relates to the choice of address format, rather than to the details of such formats.

Fixes https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r396211561

jonathanknowles committed:
Use consistent terminology for the "requested output list".

Fixes https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r396240275

jonathanknowles committed:
Clarify description of `UTxO Not Fragmented Enough`.

Fixes https://github.com/input-output-hk/cardano-coin-selection/pull/16/files/3d43e27b18a997cc4eaa61c93790ebf81b45ff04#r396240494

jonathanknowles committed:
Use consistent terminology for the "improved UTxO selection".

Fixes https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r396244102

rvl commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+
+A _transaction input_ is a pair of values (**_h_**, **_n_**), where:
+
+  * **_h_** is the hash of an existing transaction **_t_**;
```
Any reason why this is hash and not transaction id, which is more general?

You could add a definition for Transaction ID as a unique identifier that represents a transaction. Coin selection algorithms are agnostic to the format of txid, but in cardano mainnet, it's the blake2b-256 hash of the cbor-encoded transaction.

KtorZ replied:
Agree, it doesn't quite matter what `h` is here (a hash, an encrypted digest .. whatever) but what's important though is that `h` must be a **unique** transaction identifier.

jonathanknowles replied:
@rvl wrote:
> Any reason why this is hash and not transaction id, which is more general?

@KtorZ also wrote:
> it doesn't quite matter how h is calculated here (a hash, an encrypted digest .. whatever) but what's important though is that h must be a unique transaction identifier.

Thanks for this. I've fixed this in the following change: a76cc0b6201e3e52cef6c2239517558621d7502d

I've also added this to the [list of types that can be generalized](https://github.com/input-output-hk/cardano-coin-selection/issues/17) within `cardano-coin-selection`.

rvl commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+
+ 2. #### Initial UTxO Set
+
+    A subset of a wallet's [UTxO set](#utxo-set).
```
You could say "The wallet's UTxO set, or a subset of it." Then the last sentence won't be necessary.

KtorZ replied:
Or "A UTxO set to choose from", that it comes from a wallet is quite anecdotal.

jonathanknowles replied:
@rvl @KtorZ 

I've attempted to address this issue here: 4ea2f9089c135796f8fdd873a2749eff1c9b3516

Hopefully the new wording is sufficiently general, but let me know if you think it could be improved further. :+1:

rvl commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+    wallet.
+
+## Properties
+
```
I wonder if this section needs revision so that it's more human readable.

Will we also have a formal specification of coin selection?

jonathanknowles replied:
@rvl wrote:
> I wonder if this section needs revision so that it's more human readable.

Thanks, this kind of feedback is very useful.

I'm currently attempting to write each property definition so that it has _both_ a human-readable explanation (with natural language), _and_ a more mathematical definition underneath. This way, people who are not interested in mathematics will still be able to gain an intuition of each property, but those who require more detail can read the mathematical definitions.

Do you think the human-readable explanations are accessible enough?

My interpretation of the [requirements for this specification](https://jira.iohk.io/browse/ADP-189) is that someone who's never seen the Haskell code _should_ be able to read it and then be able to re-implement these algorithms (in their language of choice). Of course, with a human language explanation, it's possible that we'll miss something. But I think it's okay for us to make further revisions to the specification if people identify things that are not clear enough.

> Will we also have a formal specification of coin selection?

I'm not sure of the answer to that! @KtorZ, are you able to shed some light on this area?

jonathanknowles replied:
Closing this section after internal discussion with @rvl, as it seems we really do need to have both of the following:

* **human-language explanations**, to help casual readers gain an intuition of what each property is for.
* **mathematical laws**, to help people that want to re-implement these algorithms.

It may be the case that in future we'll write a more formal specification for coin selection, but that shouldn't necessarily restrict the way we write this document.

rvl commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+### Conservation of UTxO
+
+This property states that the [initial UTxO set](#initial-utxo-set) is
+_conserved_ in the [results](#results).
```
This seems to be a rewording of "conservation of UTxO".

Maybe say something like "This property states that every UTxO in the initial set is included in either the coin selection inputs, or in the remaining utxo set, but not both."

jonathanknowles replied:
> Maybe say something like "This property states that every UTxO in the initial set is included in either the coin selection inputs, or in the remaining utxo set, but not both."

Thanks, this wording is very clear and concise!

Fixed in e2a6a6ad2f50bc1f79b70a35b7bcfb5a82c2058b.

rvl commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+    All algorithms require that there is _at least one_ UTxO entry available
+    _for each_ requested output.
+
+  * #### UTxO Fully Depleted
```
What is the difference between fully depleted and insufficient balance?

jonathanknowles replied:
@rvl wrote:
> What is the difference between fully depleted and insufficient balance?

Good question!

If the balance of a UTxO is insufficient to pay for the requested outputs, then that is a condition that we can detect _immediately_. It's equivalent to `sumValue utxo < sumValue outputs` (assuming a suitable `sumValue` function).

However, even if the balance _appears_ to be sufficient at the outset,  coin selection algorithms can _still_ end up exhausting all of the UTxO entries, due to the fact that coin selection algorithms don't always choose UTxO entries in a way that is "optimal". One complication here is the self-imposed restriction that any given UTxO entry can only be used to pay for a single output.

@KtorZ If we do run into the `UTxOFullyDepleted` condition (which I suspect is rare), is there a good reason to not relax the above restriction, so that, if needed, we allow a UTxO entry to pay for more than one output? This could be a very simple algorithm that doesn't attempt to perform any other optimization.

jonathanknowles replied:
@rvl I've made an additional change to clarify the difference here: 327efe4e07672933777a6b775463148e3817b1bc

Let me know if you think it can be clarified further!

rvl reviewed (approved):
Nice doc - very clear.

Will there be a discussion of the tx fee being dependent on the coin selection which is dependent on tx fee? Is that what fee balancing is?

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+## What is Coin Selection?
+
+Coin selection is the process of choosing unspent coins from a wallet in order
+to pay money to one or more recipients.
```
Shouldn't the document actually starts with the definitions? For example, here we are already talking about _coins_ and _wallet_ without having really defined those terms. Unless the sense given to these terms is the common sense of a coin and wallet.

jonathanknowles replied:
> Shouldn't the document actually starts with the definitions? 

That's a good question.

My original intention was to have a brief, human-readable introduction that _doesn't_ rely on the reader to first have to read through a set of definitions.

The question I'm now considering is this: would a reader that is representative of our target audience be able to comprehend the introduction section _without_ too much difficulty, and _without_ having to jump to the definitions section?

One possible alternative solution: we could add more **hyperlinks** from terms to their definitions, as we have done in other places. Thoughts?

jonathanknowles replied:
@KtorZ I've now significantly revised both the background and definitions sections.

I've tried to keep the background section self-contained, so that it is readable by someone who hasn't necessarily read the definitions.

Though where appropriate, I've added hyperlinks from the background to the definitions, so the reader can gain greater understanding when they feel it is necessary.

Feel free to let me know if you would like to see further revisions!

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+
+    1. Having a small UTxO set limits the number of payments that we can make
+       in parallel. Since a single UTxO entry can only be used to pay for a
+       single output, we need at least as many UTxO entries as there are
```
> Since a single UTxO entry can only be used to pay for a single output

This is **not** generally true and in particular, not true for Cardano. This is a limitation our coin selection algorithm self-imposed to itself to make the problem easier to solve (so a multi-variable problem becomes multiple independant single-variable problems, easier to solve).

jonathanknowles replied:
> > Since a single UTxO entry can only be used to pay for a single output

@KtorZ wrote:
> This is **not** generally true and in particular, not true for Cardano. This is a limitation our coin selection algorithm self-imposed to itself to make the problem easier to solve (so a multi-variable problem becomes multiple independant single-variable problems, easier to solve).

**Hypothetical question:** is there a good reason that we could not introduce (in some future version of the wallet) a **_third algorithm_** (to fall back on after trying `Random-Improve` and `Largest-First`) that _relaxes_ the above restriction, and just attempts to fulfill the payment (without attempting any other optimizations)?

jonathanknowles replied:
>> Since a single UTxO entry can only be used to pay for a single output

> This is **not** generally true and in particular, not true for Cardano. 

I've made a change which fixes this here: 02711d3cd95a6d5cb446b7239732e4a95d80e2ee.

KtorZ replied:
> Hypothetical question: is there a good reason that we could not introduce (in some future version of the wallet) a third algorithm (to fall back on after trying Random-Improve and Largest-First) that relaxes the above restriction, and just attempts to fulfill the payment (without attempting any other optimizations)?

In theory, no, but there has been a fair amount of research and simulations done regarding the current algorithms. Thus introducing a third one isn't necessarily a piece of cake; now, we'd need to see the influence of a third component on the UTxO

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+
+ * A coin selection algorithm should, over the course of time, aim to generate
+   and maintain a UTxO set with _useful_ outputs: that is, outputs that allow
+   us to process future payments with a _minimum_ number of inputs.
```
I'd use `reasonably small` instead of `minimum` here :)

jonathanknowles replied:
@KtorZ wrote:
> I'd use `reasonably small` instead of `minimum` here :)

The "minimum number of inputs" wording actually appears in [Self Organisation in Coin Selection](https://iohk.io/en/blog/posts/2018/07/03/self-organisation-in-coin-selection/).

However, I think "reasonably small" also works well here.

Fixed in e914b068a57c76fe423545eca5c8ad315a05a068.

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+   us to process future payments with a _minimum_ number of inputs.
+
+ * A coin selection algorithm should employ strategies to limit the
+   amount of _dust_ that accumulates in the UTxO set.
```
Isn't this second property redundant with the first one?

jonathanknowles replied:
@KtorZ wrote:
> Isn't this second property redundant with the first one?

I'm not sure that it is redundant.

In theory, it seems possible to have a UTxO that simultaneously has _both_ a lot of dust _and_ a lot of useful outputs.

Though perhaps it would make sense to change the order of these two properties around?

We could write:

> A coin selection algorithm should:
> * employ strategies to limit the amount of _dust_ that accumulates in the UTxO set.
> * over the course of time, aim to generate and maintain a UTxO set with _useful_ outputs: that is, outputs that allow us to process future payments with a _reasonably small_ number of inputs.

Thoughts?

jonathanknowles replied:
I've made the modification indicated above here https://github.com/input-output-hk/cardano-coin-selection/pull/16/commits/31bf5407abf07e6a015f2d5e48f47994dc80a5c3.

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+
+An _address_ is a unique identifier that represents a destination for a payment.
+
+Addresses are owned (and generated) by individual wallets.
```
*are typically owned 

In practice, HD wallets aren't necessary to generate addresses and interact with the blockchain.

jonathanknowles replied:
> *are typically owned
> 
> In practice, HD wallets aren't necessary to generate addresses and interact with the blockchain.

Good point!

Fixed in 786e107d21138b5ab9797f486ac2bc96f159f405.

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+
+### Coin Amount
+
+A _coin amount_ is a non-negative integer value that represents a number of
```
It even is a _positive_ integer value. There's in principle, no such thing as a coin of value `0` (although quite practical in many calculations). This is true at least in Byron and I know there was discussion for Shelley about whether or not a transaction with an output of value `0` should be considered valid.

jonathanknowles replied:
> It even is a _positive_ integer value. There's in principle, no such thing as a coin of value `0` (although quite practical in many calculations). This is true at least in Byron and I know there was discussion for Shelley about whether or not a transaction with an output of value `0` should be considered valid.

Good catch. I initially wanted to include the possibility of zero, as this would allow us to talk about temporary amounts used in calculations (which occasionally end up as `0`, only to be filtered out later on). On reflection, I think it makes sense to exclude this case.

Fixed in 64e0dfc49e88f2ff1c3bd93a279b3a44d0d9ceec.

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+
+Each member of the set is a pair of the form (**_i_**, **_o_**), where:
+
+  * **_i_** is a [transaction input](#transaction-input);
```
A UTxO is actually a transaction identifier and an output index. I am not sure to agree with the defnition here :)

KtorZ replied:
(see also: https://github.com/input-output-hk/cardano-wallet/wiki/About-UTxO )

jonathanknowles replied:
@KtorZ wrote:
> A UTxO is actually a transaction identifier and an output index. I am not sure to agree with the definition here

Fair enough!

When I originally wrote this article, I was attempting to be consistent with the types defined in our code, which _effectively_ defines the UTxO set as a set of pairs of (`TxIn`, `TxOut`):

```hs
newtype UTxO = UTxO { getUTxO :: Map TxIn TxOut }
```

However, as far as coin selection algorithms are concerned, the `UTxO` could also be defined as:

```hs
newtype UTxO k = UTxO { getUTxO :: Map k Coin }
```

From the point of view of a coin selection algorithm, so long as `k` is instantiated to a type that provides an `Ord` instance, the other details are irrelevant. (See https://github.com/input-output-hk/cardano-coin-selection/pull/29.)

The question then becomes:
> How can we relate this type to the human-readable definition that we provide?

We could write something like this:

> A _UTxO set_, or the set of _unspent transaction outputs_, represents the unspent value associated with a wallet.
>
> From the point of view of a coin selection algorithm, each entry in the UTxO set consists of a pair of the form (**_o_**, **_v_**), where:
>   * **_o_** is a [transaction output](#transaction-output);
>   * **_v_** is a [coin value](#coin-value).

Which can then reuse our _existing_ definition of _transaction output_:
> a pair (**_h_**, **_n_**), where:
>  * **_h_** is a _unique identifier_ for an existing transaction _t_;
 > * **_n_** is a 0-based integer index into the output list of transaction _t_.

Thoughts?

jonathanknowles replied:
@KtorZ I've now completely revised this section with new definitions.

See: https://github.com/input-output-hk/cardano-coin-selection/pull/16/commits/4f266317a99ca3ccf87f1c98a302bb3f617f2a25

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+A given [transaction input](#transaction-input) can only appear _once_ in a
+given UTxO set.
+
+### Transaction Input
```
Note that a transaction input is **by definition**, a UTxO.

jonathanknowles replied:
See discussion here: https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r400068972

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+
+    If a coin selection algorithm is applied to the _complete_ UTxO set of a
+    wallet, then the remaining UTxO set represents the _updated_ UTxO set of the
+    wallet.
```
I wouldn't put too much emphasis on wallets in this document. A coin selection selects elements from a UTxO set. Period. This can obviously be applied to UTxO management inside a wallet, but that is out-of-scope of the current spec IMO.

jonathanknowles replied:
@KtorZ wrote:
> I wouldn't put too much emphasis on wallets in this document. 

I agree that we should avoid making the descriptions of these algorithms depend on the details of wallets. Though perhaps there is a balance that we can strike here?

Of course, the concept of a coin selection algorithm can be completely divorced from the concept of a wallet. So we take a lot of care, for example, to talk about the `initial UTxO set` rather than the `wallet UTxO set`.

But perhaps it helps to occasionally mention the concept of a wallet, as that is a very common context in which a UTxO appears, and it gives readers a bit of context for how these algorithms are used.

How about the following **alternative wording**:

> In the context of a _wallet_, if a coin selection algorithm is applied to the wallet's _complete_ UTxO set, then the remaining UTxO set represents the _updated_ UTxO set of that wallet.

jonathanknowles replied:
@KtorZ I've now updated this section with the alternative wording indicated above.

See here: 874538b1e28327b8004d9eaef83113d159c019e6.

jonathanknowles commented:
@rvl wrote:
> Will there be a discussion of the tx fee being dependent on the coin selection which is dependent on tx fee? Is that what fee balancing is?

Yes! There will definitely be a discussion about fee balancing.

At the moment, I'm planning to put this in the next draft. Will let you know when it's ready for review. :+1:

jonathanknowles committed:
Change definition of "Coin Amount" to a "positive integer value".

Addresses https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r397877789

jonathanknowles committed:
Adjust wording to reflect that addresses are /typically/ owned by wallets.

This is isn't necessarily always the case however, as addresses can be
generated outside the context of a wallet. So we adjust the wording to
reflect this case.

Fixes: https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r397876465

jonathanknowles committed:
Revise definition of `Initial UTxO Set` to make it more general.

A UTxO set doesn't necessarily have to be associated with a wallet, so
we adjust the definition to reflect this fact.

Addresses the following feedback:

https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r397775347
https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r397882094

jonathanknowles committed:
Simplify definition of "Conservation of UTxO" property.

Addresses: https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r397775769

jonathanknowles committed:
Replace "minimum" with "reasonably small".

In response to feedback:

https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r397875312

jonathanknowles committed:
Add justification for giving priority to the Random-Improve algorithm.

Addresses the following feedback:

https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r396241386

jonathanknowles commented on information/specification.md:
```diff
@@ -0,0 +1,706 @@
+# External Resources
+
+## Self Organization in Coin Selection
+
```
Suggestion: add a short summary of the content of this resource, to explain to readers why they should want to read it.

jonathanknowles replied:
@jonathanknowles wrote:
> Suggestion: add a short summary of the content of this resource, to explain to readers why they should want to read it.

Fixed in b778151bdb1eed088fbf83be66b210f43c9b066c.

jonathanknowles committed:
Improve the description of cardinality restrictions.

There is no general principle that requires coin selection algorithms to
ensure that any given UTxO entry is used to pay for at most one output.

It is the case that both the Largest-First and Random-Improve algorithms
impose this restriction, but this is done in order to reduce the
complexity of the optimisation problems they attempt to solve.

In principle, we could have an algorithm that relaxes this restriction.
This would make it possible (for example) to pay for multiple smaller outputs
with a single large UTxO entry.

Since there is no general principle that requires a restriction, this
change removes the "Cardinality" subsection from the common "Interface"
section, and creates equivalent "Cardinality" subsections within the
descriptions of the "Largest-First" and "Random-Improve" algorithms.

This change also updates the "Failure Modes" section accordingly.

jonathanknowles committed:
Add an introduction to the "Self Organization in Coin Selection" resource.

jonathanknowles committed:
Remove the requirement for a transaction identifier to be a hash value.

Addresses the following feedback:

https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r397775176
https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r397880078

Anviking commented on information/specification.md:
```diff
@@ -0,0 +1,733 @@
+to pay money to one or more recipients.
+
+In the world of physical money, wallets hold value in the form of _coins and
+banknotes_.  When making a payment, we typically select a number of coins and
```
The "typically" here makes me think of: "what is the alternative to selecting a number of coins and banknotes", and "is it true that I / one typically does this?", which I think distracts me from the point.

I wonder if it wouldn't be more effective to use a specific concrete example instead of having to _correctly_ generalise the typical scenario.

E.g, rather than
```
When making a payment [...] 
we typically select [...] 
it's often impossible to select the exact amount [...] "
in such cases we typically give the recipient more
```
what about something more like
```
Alice buys food [...] 
she selects [...] 
she cannot select the exact amount [...]
she pays more than"
```
?

jonathanknowles replied:
> I wonder if it wouldn't be more effective to use a specific concrete example instead of having to correctly generalise the typical scenario.

Perhaps we could provide both? Some people do seem to prefer general definitions, but it's clear that many people also like examples.

jonathanknowles replied:
Fixed in https://github.com/input-output-hk/cardano-coin-selection/pull/16/commits/638cd7749ea454037c0f835be180bbd654f73cdd.

jonathanknowles committed:
Add example of coin selection in the physical world.

jonathanknowles committed:
Add emphasis.

jonathanknowles committed:
Revise the `Coin Selection in Cardano` section.

The aim of this revision is to make this section more self-explanatory,
without requiring the reader to have already read through the
definitions section. (Though they may read definitions by following
hyperlinks.)

jonathanknowles committed:
Rename `Coin Amount` to `Coin Value`.

jonathanknowles committed:
Add hyperlink to `UTxO Set`.

jonathanknowles committed:
Simplify the goals section.

jonathanknowles committed:
Revise definitions section.

jonathanknowles committed:
Simplify discussion of change in background section.

jonathanknowles committed:
Add definition for dust output.

jonathanknowles committed:
Add definition for change output.

jonathanknowles committed:
Add more links to the "Why is Coin Selection Non-Trivial?" section.

jonathanknowles committed:
Move dust output and change output definitions to after UTxO.

jonathanknowles committed:
Add entries to the table of contents.

piotr-iohk commented on information/specification.md:
```diff
@@ -0,0 +1,826 @@
+
+There are a number of ways in which a coin selection algorithm can fail:
+
+  * #### UTxO Balance Insufficient
```
Does "Not enough money to cover fee" fall into that category, or maybe it could be mentioned as separate one? (Just wondering... :thinking: )

jonathanknowles replied:
Hi @piotr-iohk 
> Does "Not enough money to cover fee" fall into that category, or maybe it could be mentioned as separate one? (Just wondering... thinking )

That's a good point!

I've not yet added the section on fee balancing. But that will include the error you mentioned (as it falls into a separate category of errors).

piotr-iohk reviewed (approved):
:100:

jonathanknowles committed:
Revise section on applying a coin selection algorithm to a wallet.

Of course, the concept of a coin selection algorithm can be completely
divorced from the concept of a wallet. So we take a lot of care, for
example, to talk about the /initiall UTxO set rather than the /wallet/ UTxO
set.

But occasionally placing coin selection in the context of a wallet is
helpful, as that is a very common context in which a UTxO appears, and
it gives readers a bit of context for how these algorithms are used.

Response to feedback here:

https://github.com/input-output-hk/cardano-coin-selection/pull/16#discussion_r397885226

jonathanknowles committed:
Revise the section on experimental evidence.

Explicitly naming the "Self Organization in Coin Selection" resource makes
this section a bit clearer.

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,826 @@
+    This failure occurs if the _number_ of UTxO entries needed to pay for the
+    outputs in the [requested output list](#requested-output-list) exceeds the
+    upper limit specified by the [maximum input count](#maximum-input-count)
+    parameter.
```
The maximum input count parameter is more of a symptom IMO. This doesn't really tell you why it is wrong to select too many entries (although mentioned in previous sections). What about:

"This failure occurs when another input must be selected by the algorithm to continue but doing so will increase the size of the resulting transaction beyond an acceptable limit."

jonathanknowles replied:
> This failure occurs when another input must be selected by the algorithm to continue but doing so will increase the size of the resulting transaction beyond an acceptable limit.

Thanks. I've incorporated this suggestion into the following change: e31e212078ded1f1b59804f245d4d8c6683b07b5

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,826 @@
+In particular, the _outputs_ field of the [coin selection](#coin-selection)
+result should be _equal to_ the [requested output list](#requested-output-list).
+
+## Failure Modes
```
It feels odd to have both the possible failures and the properties described before even talking about the algorithms themselves. What about changing the order of the sections here so that we first describe the algorithms, and then, talk about the properties they have. 

It's like going through the troubleshooting section of an installation manual directly without even having looked at the installation instructions in the first place.

jonathanknowles replied:
> It feels odd to have both the possible failures and the properties described before even talking about the algorithms themselves. What about changing the order of the sections here so that we first describe the algorithms, and then, talk about the properties they have.

Hmm. I'm not sure I agree here.

I would argue that the set of potential failure modes is actually part of the API, as it represents a set of errors common to all algorithms. The API itself defines the set of potential errors. (This is also the case in the Haskell library.)

Similarly, the properties described in the properties section apply to all possible algorithms (and not just those we have provided). They specify additional restrictions on the relationships between parameters and results that are not practical to express in the types themselves.

The separation between API and implementation actually mirrors the separation we have in the library itself.

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,826 @@
+value of that output.
+
+The name of the algorithm is taken from the idea that the **largest** UTxO
+entry is always considered **first**.
```
It'll be good to give a disclaimer here about why is Largest First used as (and only used at) a fallback. There are two main reasons:

a) If we fail to cover a payment when selecting coins from the largest first, then **necessarily**, it is impossible to cover such payment with the given UTxO set.  <-- and this is an actual property of the Largest First algorithm.

b) Always selecting largest coins first to pay for transaction has dramatic effects on the UTxO in the long run for it tends to merge large UTxOs together (reducing payment concurrency) and produces dust (increasing cost of making transactions and possibly making wallets unusable). It is thereby not a viable option to use on a typical use-case.

Also about a), note that our _current implementation_ has an important flaw: it does select largest coins first, but consider each output independently instead of considering the total balance. This is **wrong**. To be honest I thought we fixed that but in doubt, I double-checked and, we actually haven't :no_mouth:

jonathanknowles replied:
@KtorZ wrote:
> Also about a), note that our current implementation has an important flaw: it does select largest coins first, but consider each output independently instead of considering the total balance. This is wrong. To be honest I thought we fixed that but in doubt, I double-checked and, we actually haven't :no_mouth: 

I've created a bug report for this here: https://github.com/input-output-hk/cardano-coin-selection/issues/39

jonathanknowles replied:
Fixed in d35f56c06f8924c403b915317e967ea84d212801.

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,826 @@
+
+### Cardinality
+
+The Largest-First algorithm imposes the following cardinality restriction:
```
See my previous comment, this restriction should only apply to the Random-Improve algorithm!

jonathanknowles replied:
Fixed in d35f56c06f8924c403b915317e967ea84d212801.

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,826 @@
+outputs](#requested-output-list) in _descending order of coin value_, from
+_largest_ to _smallest_. For each output, the algorithm repeatedly selects
+entries at **random** from the [initial UTxO set](#initial-utxo-set), until the
+_total value_ of selected entries is enough to pay for that ouput.
```
for that _ouput_ → for that _ou**t**put_

jonathanknowles replied:
> for that _ouput_ → for that _ou**t**put_

Fixed in e252dd3c33f9a076c2a08a3ed0d74a494c491880.

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,826 @@
+
+In the _second_ phase, the algorithm attempts to **improve** upon each of the
+UTxO selections made in the previous phase, by conservatively expanding the
+selection made for each output, in order to generate improved change values.
```
It'll be worth mentioning here what a "improved" output is and why is that so:

- A new change value is considered an improvement if its value gets _closer_ to the value of the corresponding output being covered in the first place. 

The algorithm works by considering, for each output, the triple (minValue, idealValue, maxValue) of values to cover. `minValue` is obviously the value of the selected output itself and, to be considered successful, the algorithm **must* select inputs to cover for at least this `minValue`. And then, improve to try covering for `idealValue`, but no more than `maxValue` (otherwise, we would end up selecting a lot more than what's really needed). In practice, `idealValue` is set to twice the output value and `maxValue` to thrice the output value. 
This therefore aims for constructing change addresses that are roughly equal to the output value (with an error margin or +/- 1 output value; which turns out to be good enough in simulations and empirical observations). 
On large UTxO sets where the variety of coins is bigger, one could imagine using more restrictive constraints like:

- min = 1.75 * output
- ideal = 2*output 
- max = 2.25 * output

such that change outputs are roughly of the same size as the output but with a greater precision.

jonathanknowles replied:
@KtorZ I've extended the description of the algorithm in 4cf7b1334a40d382f2642ca1187a9f12a339f8d6.

The above change attempts to explain the motivation behind attempting to produce a selection for each output that's around twice the value of that output.

Let me know if you think it can be improved further. (I'll also make changes to subsequent sections, where appropriate.)

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,826 @@
+In this phase, the algorithm attempts to improve upon each of the UTxO
+selections made in the previous phase, by conservatively expanding the
+selection made for each output in order to generate improved change
+values.
```
It could be nice to point out that this phase can not make the algorithm fails and is totally optional. As per the 3rd principle, failing to improve _probably_ means that there's no need to improve in the first place which is part of the beauty of this whole algorithm.

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,826 @@
+ 1.  **Calculates a _target range_** for the total value of inputs used to
+     pay for that output, defined by the triplet:
+
+     (_minimum_, _ideal_, _maximum_) = (_v_, _2v_, _3v_)
```
Okay so this is mentioned here. I think it'll deserve a small mention in the overview above, at least to explain the goal of the improvement step: getting change values closer to the output value.

jonathanknowles replied:
Done! See 4cf7b1334a40d382f2642ca1187a9f12a339f8d6.

KtorZ commented on information/specification.md:
```diff
@@ -0,0 +1,826 @@
+If for each payment request of value **v** we create a change output of
+_roughly_ the same value **v**, then we will end up with a distribution of
+change values that matches the typical value distribution of payment
+requests.
```
This one maybe deserves a small example? 

"To give a concrete example: if Alice often buys bread for 1€, then her wallet will self-organize itself over time to contain a lot of 1€ coins."

jonathanknowles replied:
> This one maybe deserves a small example?
> 
> "To give a concrete example: if Alice often buys bread for 1€, then her wallet will self-organize itself over time to contain a lot of 1€ coins."

Done! Added in 3471486a2031a36a47d9dd42d12aad897c4027e2.

rvl reviewed (approved):
:star: Great doc, this revision is even better than the previous.

jonathanknowles commented on information/specification.md:
```diff
@@ -0,0 +1,826 @@
+    the _updated_ UTxO set of that wallet.
+
+## Properties
+
```
We should probably insert a short paragraph here to explain what this section is about.

jonathanknowles replied:
> We should probably insert a short paragraph here to explain what this section is about.

Fixed in e82a7641abf2c09bdb967a7558e89a6ab196111c.

jonathanknowles committed:
Add paragraph to explain Properties section.

jonathanknowles committed:
Correct various spelling errors and inconsistencies.

jonathanknowles commented on information/specification.md:
```diff
@@ -0,0 +1,830 @@
+
+### Coin Value
+
+A _coin value_ is a positive integer value that represents a number of
```
This should be changed back to **_non-negative_**.

See [this comment](https://github.com/input-output-hk/cardano-coin-selection/issues/38#issuecomment-618262828) on issue https://github.com/input-output-hk/cardano-coin-selection/issues/38.

jonathanknowles committed:
Break up long paragraphs.

jonathanknowles committed:
Rework the "Why is Coin Selection Non-Trivial?" section.

jonathanknowles committed:
More paragraph splitting.

jonathanknowles committed:
Linkify transaction size limitation.

jonathanknowles committed:
Add emphasis.

jonathanknowles committed:
Clarify what is meant by a dust output.

jonathanknowles committed:
Add an introduction to the definitions section.

jonathanknowles committed:
Avoid linebreak within link text.

jonathanknowles committed:
Expand the "Interface" introduction.

jonathanknowles committed:
Use "All coin selection functions" wording.

jonathanknowles committed:
Emphasise that a coin selection is the basis for a transaction.

jonathanknowles committed:
Revise wording of "Interface" section.

jonathanknowles committed:
Clarify the definition of a coin selection's inputs.

jonathanknowles committed:
Some simplifications to wording.

jonathanknowles committed:
Clarify the definition of an address.

jonathanknowles committed:
Revise the definition of coin value.

Clarify that a coin must be non-negative.

jonathanknowles committed:
Clarify section on transaction input.

jonathanknowles committed:
Add hyperlinks to "coin value" and "transaction".

jonathanknowles committed:
Clarify definition of UTxO set.

jonathanknowles committed:
Clarify definition of change output.

jonathanknowles committed:
Clarify Interface section.

jonathanknowles committed:
Clarify the definition of coin selection.

jonathanknowles committed:
Revise "Coverage of Payments" property.

jonathanknowles committed:
Revise "Conservation of UTxO".

jonathanknowles committed:
Make `Largest-First` pay for outputs collectively.

Fixes:

https://github.com/input-output-hk/cardano-coin-selection/issues/39

jonathanknowles committed:
Clarify introduction to "Largest-First".

jonathanknowles committed:
Add maximum input count condition to Largest-First.

jonathanknowles committed:
Clarify largest-first condition.

jonathanknowles committed:
Shorten UTxO variable names.

jonathanknowles committed:
Clarify variable names in Largest-First.

jonathanknowles committed:
Simplify the result description for "Largest-First".

jonathanknowles committed:
Use "requested output set" instead of "requested output list".

jonathanknowles committed:
Use bold, italics, and subscript consistently for variables.

jonathanknowles committed:
Replace well-ordered with totally-ordered.

jonathanknowles committed:
Revise Conservation of UTxO laws.

jonathanknowles committed:
Add "and" clause.

jonathanknowles committed:
Indent mathematical law.

jonathanknowles committed:
Use `n` to represent counts of things.

jonathanknowles committed:
Remove draft header message.

jonathanknowles committed:
Add draft CIP header.

jonathanknowles committed:
Add abstract.

jonathanknowles committed:
Use present tense instead of future tense.

jonathanknowles committed:
Add example icon.

jonathanknowles committed:
Describe the reasoning behind the maximum input count.

jonathanknowles committed:
Add Haskell reference implementations.

jonathanknowles committed:
Expand the introduction to Random-Improve.

Explain the motivation behind choosing a UTxO selection whose value is
approximately twice the value of the output to which it corresponds.

jonathanknowles committed:
Remove redundant section.

jonathanknowles committed:
Add an example to the Change Management section.

jonathanknowles committed:
Rename section to "Goals of Coin Selection Algorithms".

jonathanknowles committed:
Remove blank space.

jonathanknowles committed:
Add missing word "algorithm".

jonathanknowles committed:
Emphasize "additional" entries.

jonathanknowles committed:
Replace "degree" with "extent".

jonathanknowles committed:
Further simplify the introduction to "Random-Improve".

jonathanknowles committed:
Clarify that the change output value is a target value.

jonathanknowles committed:
Use "change coins" wording in example.
